File: ecommerce-api-legacy/src/middlewares/errorHandler.js
class AppError extends Error {
  constructor(message, statusCode = 500) {
    super(message);
    this.statusCode = statusCode;
  }
}

class NotFoundError extends AppError {
  constructor(message) { super(message, 404); }
}

class ValidationError extends AppError {
  constructor(message) { super(message, 400); }
}

class PaymentDeniedError extends AppError {
  constructor(message = 'Pagamento recusado') { super(message, 400); }
}

function errorHandler(err, req, res, _next) {
  const status = err.statusCode || 500;
  if (status >= 500) console.error(err);
  res.status(status).json({ error: err.message || 'Internal server error' });
}

module.exports = { AppError, NotFoundError, ValidationError, PaymentDeniedError, errorHandler };

File: task-manager-api/src/middlewares/error_handler.py
import logging

from flask import jsonify
from werkzeug.exceptions import HTTPException

log = logging.getLogger(__name__)


class AppError(Exception):
    status_code = 500

    def __init__(self, message, status_code=None):
        super().__init__(message)
        if status_code is not None:
            self.status_code = status_code


class BadRequestError(AppError):
    status_code = 400


class UnauthorizedError(AppError):
    status_code = 401


class ForbiddenError(AppError):
    status_code = 403


class NotFoundError(AppError):
    status_code = 404


class ConflictError(AppError):
    status_code = 409


def register_error_handlers(app):
    @app.errorhandler(AppError)
    def _handle_app_error(err):
        return jsonify({"error": str(err)}), err.status_code

    @app.errorhandler(HTTPException)
    def _handle_http_exception(err):
        return jsonify({"error": err.description}), err.code

    @app.errorhandler(Exception)
    def _handle_unexpected(err):
        log.exception("Unexpected error")
        return jsonify({"error": "Internal server error"}), 500
